eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -140,264 +506,302 @@ proposalsAddVote voter vote gai (Proposals omap) =
+              )
+
+-- | Get all the descendents of an action-id from the @`Proposals`@ forest
+getAllDescendents ::
+  forall era.
+  Proposals era ->
+  GovActionId (EraCrypto era) ->
+  Set (GovActionId (EraCrypto era))
+getAllDescendents (Proposals omap _roots hierarchy) gai = case OMap.lookup gai omap of
+  Nothing -> Set.empty
+  Just gas -> case gas ^. gasActionL of
+    ParameterChange _parent _ _ -> collectDescendents pfhPParamUpdateL
+    HardForkInitiation _parent _ -> collectDescendents pfhHardForkL
+    TreasuryWithdrawals _ _ -> Set.empty
+    NoConfidence _parent -> collectDescendents pfhCommitteeL
+    UpdateCommittee _parent _ _ _ -> collectDescendents pfhCommitteeL
+    NewConstitution _parent _ -> collectDescendents pfhConstitutionL
+    InfoAction -> Set.empty
+  where
+    collectDescendents ::
+      Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+      Set (GovActionId (EraCrypto era))
+    collectDescendents lenz = Set.map unGovPurposeId $ go lenz $ GovPurposeId gai
+    go ::
+      forall p.
+      Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+      GovPurposeId p era ->
+      Set (GovPurposeId p era)
+    go lenz gpi =
+      case Map.lookup gpi $ hierarchy ^. lenz . pHierarchyNodesL of
+        Nothing -> error "Impossible! getAllDescendents: GovPurposeId not found"
+        Just (PNode _parent children) -> children <> foldMap (go lenz) children
+
+-- | Remove the set of given action-ids with their descendents from the
+-- @`Proposals`@ forest
 proposalsRemoveDescendentIds ::
   Set (GovActionId (EraCrypto era)) ->
   Proposals era ->
   (Proposals era, Map (GovActionId (EraCrypto era)) (GovActionState era))
-proposalsRemoveDescendentIds gais ps@(Proposals omap) =
-  proposalsRemoveIds (gais <> foldMap (getAllDescendents omap) gais) ps
+proposalsRemoveDescendentIds gais ps =
+  proposalsRemoveIds (gais <> foldMap (getAllDescendents ps) gais) ps
+
+-- | For use in the @`EPOCH`@ rule. Apply the result of
+-- @`extractDRepPulsingState`@ to the @`Proposals`@ forest, so that:
+--   i. all the expired action-ids and their descendents are removed,
+--   and
+--   ii. the sequence of enacted action-ids is promoted to the root,
+--   removing competing/sibling action-ids and their descendents at each
+--   step
+proposalsApplyEnactment ::
+  forall era.
+  Seq (GovActionId (EraCrypto era)) ->
+  Set (GovActionId (EraCrypto era)) ->
+  Proposals era ->
+  ( Proposals era
+  , Map (GovActionId (EraCrypto era)) (GovActionState era) -- Removed due to enactment
+  , Map (GovActionId (EraCrypto era)) (GovActionState era) -- Removed due to expiry
+  )
+proposalsApplyEnactment enactedGais expiredGais props =
+  let (unexpiredProposals, expiredRemoved) = proposalsRemoveDescendentIds expiredGais props
+      (enactedProposalsState, enactedRemoved) = foldl' enact (unexpiredProposals, Map.empty) enactedGais
```

Comment:
Let's err on the side of caution and return the original. This way nothing will be removed. This is what we should do:
```suggestion
      (enactedProposalsState, enactedRemoved) =
        case foldlM' enact (unexpiredProposals, Map.empty) enactedGais of
          Nothing -> assert False (props, mempty, mempty)
          Just res -> res
```